### PR TITLE
Allow authors (and users) to customize array option's entry names

### PR DESCRIPTION
**Pull request by emptyrivers**
# Description

This is only approximately 80% complete, and needs some heavy testing of edge cases.

This enhancement adds a new dropdown to `Author Mode` of array type options, "Name Source". This dropdown controls how the page selector in `User Mode` displays each entry in the array.

Possible Values:

- Fixed Names
	- Only available on arrays with a fixed size. In this mode, the author is expected to specify the name of each entry manually.
- Entry Order
	- This provides the default behavior, where each entry is names "Entry N".
- Any key of a suitable sub option.
	- There may be 0 or more of these. Any Range, Number, or String type option is a suitable name source. In this mode, the chosen option's user value will be used to name the entry.

### CODE REVIEW

File: WeakAurasOptions/AuthorOptions.lua [hunk 1]
```diff
@@ -1100,9 +1112,94 @@ typeControlAdders = {
...
+            WeakAuras.Add(childData)
+          end
+          WeakAuras.ReloadTriggerOptions(data)
+        end,
+      }
+      if option.nameSource == -1 then
+        for i = 1, option.size do
+          args[prefix .. "entry" .. i .. "name"] = {
+            type = "input",
+            name = nameArray(option, "entryNames", i, L["Entry %i"]:format(i)),
```
Comment (emptyrivers) [inline WeakAurasOptions/AuthorOptions.lua]: These helper functions aren't yet implemented, which will probably cause the CI check to fail.

**Comment by mrbuds** [comment]
I love this PR :)
So far i only found a tiny issue with the layout of reordering buttons next to the dropdown, i guess the spacing logic is inverted.
With "Disallow Entry Reordering" enable: https://i.imgur.com/Fawx2Ej.png
With it disable: https://i.imgur.com/mlTujWV.png

**Comment by emptyrivers** [comment]
Should be fixed now.

**Comment by InfusOnWoW** [comment]
It's very impressive how flexible you made the naming system.

File: WeakAurasOptions/AuthorOptions.lua [hunk 2]
```diff
@@ -1100,9 +1174,109 @@ typeControlAdders = {
...
+        end,
+        set = function(_, value)
+          for id, optionData in pairs(option.references) do
+            local childOption = optionData.options[optionData.index]
+            local childData = optionData.data
+            if (value == -1) ~= (childOption.nameSource == -1) then
+              if value == -1 then
+                local entryNames = {}
+                for i = 1, #childOption.subOptions do
+                  entryNames[i] = L["Entry #i"]:format(i)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AuthorOptions.lua]: This code didn't work for me.

That is, with

!nsvZYTTnq4NMESDKuStDp6FKt0mkuUc0DC7bidsUKe1qaSaGYw(GF27UlOKODoKmdhcYf7)733kNkZLITsrf9GF2yCpllKIxwvxhGOmBIuSF03kBzRZFNtBJsX1ZZYNVwkkDotL7zR4zDhiV6KG5vna5nDPZsYJEDtd4dyKY4ipiG(Sck6RRZ3JEq815lV727xI3Z)Q69QzsrOdmMfvK1yWd9fWoWgfOv6xKInxFPiFJi)Y1OZ6TAuLoJAp6CHvTfgSInbdaOmX2to5opWorC38Llj96TNsTijqvg17G8KWV5Qaz2VoDcDd2kaVvz(lSW0yDMn7mP4ja6Uet4Y4Avu7OMGYQ3sFBPQnev(ix2yTrc3Kk1aGTkSehQCRZcum2Q02FETR1wDO9Nw)i18dQiPlpUiqW1oJZNMt0Zm(9N43NrVttG1qd69rZt4L4MqRch(pmc1me9(c6(bTsF(lHHFl)XXJ1R2rZVB96xF7X)SxvH5S6ThZFFShN8tyNnHD2Kp4SK2)94mLfdM6bi(L3NVAq6oDqxyy89XmjbxxDF(YfzZhK)okYISS5R3C1Q88vFB9IV818duOVlO)BFiQR3FIv5DrCEKZTPSvK7tw(WhTKseH(veroDwAC2c6MwuJpJirJtXmBcmYK9EtuFavxAuHW3jnqo7dcjkqFedpJ9PEGRC3be)ftMn58eUj1rk6JrNDfQasbhUOvx(KfWWHdawY4cXZaPKX82cCxsfYq5sOpaNGKfjUyc1DKhng1ZCsD8yn(Jrw9AA)35MLFPR4HfU)rkE152Yjgf4XZkLPRvXLWHLC0e6gDqHyJko76JoQbrwkkvgG1MIqg8mXbpcUomPR94ckCxicLzDtZ5BjPPoI461ZNNLC5TE4)6bB5EmD(TzNt5HTw3qndxh1xMEGngPILp(u64S0X5PJpNo(90XfPJ)iDGB2oo21wAoddLgMyR6IVJ0J86tIIhy7nExFhp7UbcLK10AyHR3xcCNCyapL1bNTgvxGrwpb7pvlSvifJ)9ThrbT6kyn48v4YzuBJEREGhIRWXrqkYjjkVNWFLh8EXa4oJxDhtD0LGTHZKjCZSspQw471iSL6q8FxLICbtuKIBGDsXo04ZMrqj5))

switching to Fixed Names didn't fill in the entryNames as this code appears to do

Comment (InfusOnWoW) [reply WeakAurasOptions/AuthorOptions.lua]: Apparently that's because there's no sub option, but even with a subOption, that looks incorrect in two ways:

It should use childOption.size, it should use %i in the string

**Comment by InfusOnWoW** [comment]
Apart from that minor bug, I didn't find anything odd in my testing and while reviewing the code.

Although I didn't try to understand each line.

**Comment by InfusOnWoW** [comment]
As far as I'm concerned this is ready to go after the bug fix.
